### ISSUE TITLE
Make twisted/test/test_text pass under Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5830]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)| @vperic reported|
|-|-|
|Trac ID|trac#5830|
|Type|task|
|Created|2012-08-03 11:34:04Z|
|Branch|https://github.com/twisted/twisted/tree/test_text-5830|

As before, to actually make this actually run you need to import unittest from the library, not from trial, and then "python3.2 -m unittest twisted.test.test_text" will run the tests.

The changes:

Replace calls to functions from the string module in favor of string
methods (Refs: [#4999](https://github.com/twisted/twisted/issues/4999)); don't use the types module in favor of the
"isinstance" function - types was changed in Python 3 to not include
some base types, like dicts, lists and tuples.

This also required the addition of some tests.

Attachments:

* [test_text.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2bd/2bdbeeec1af1608ecb285345a3c34a3bfc31c4bd/89fb87bcf455a3589dbf947bcb95ca60603bc27b.patch) (5119 bytes) - added by vperic on 2012-08-03 11:34:47Z - First version
* [test_text.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2bd/2bdbeeec1af1608ecb285345a3c34a3bfc31c4bd/d47a47d7965b39e4c9d8c14236cbae40b47e6f9f.patch) (5123 bytes) - added by vperic on 2012-08-04 09:55:21Z - Second version

<details><summary>Searchable metadata</summary>

```
trac-id__5830 5830
type__task task
reporter__vperic vperic
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_test_text_5830 branches/test_text-5830
branch_author__vperic vperic
status__closed closed
resolution__fixed fixed
component__core core
keywords__py3k py3k
time__1343993644000000 1343993644000000
changetime__1344931432000000 1344931432000000
version__None None
owner__vperic vperic
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_2"></a>|@pitrou commented|
|-|-|

Changing exact type tests to isinstance() will change semantics: what happens e.g. to a tuple subclass?

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_3"></a>|@vperic commented|
|-|-|

As in, a tuple subclass would be caught now but not before? I guess you are right, but I could argue that the new behaviour is more correct. Right now, subclasses of those types are handled just right any other element, so printed just as a string, which is not what they're supposed to do per the doctest of stringyString; with the new code, they'll also be printed "one element per line".

The function itself isn't used anywhere, so I'm not sure what the original author intended.

Still, if I was to keep the old semantics, how should I fix this? The types module is changed in Python 3.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_4"></a>|@pitrou commented|
|-|-|

Replying to [vperic](#note_3):
> 
> Still, if I was to keep the old semantics, how should I fix this?

`type(foo) is tuple`

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_5"></a>|@vperic set owner to @vperic|
|-|-|

D'oh, you're right. Thanks! I'll update the patch to do just that; dropping the review tag for now.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_6"></a>|@vperic removed owner|
|-|-|

Right, fixed that, thanks antoine! Putting back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

Could you create a branch for the patch so we can run it on buildbot. This should rule out any test failures. Maybe once [#4999](https://github.com/twisted/twisted/issues/4999) landed?

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_8"></a>|@vperic set owner to @vperic|
|-|-|

Yes, definitely, I'll update this once [#4999](https://github.com/twisted/twisted/issues/4999) is in to contain just the py3k changes (which are minimal, really).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

[#4999](https://github.com/twisted/twisted/issues/4999) was merged, so this can now be updated.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_10"></a>|@vperic commented|
|-|-|

(In [35184]) Create branch test_text-5830

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_11"></a>|@vperic commented|
|-|-|

I've ordered a test run on just twisted.test.test_text (as it's the only touched module). After removing the added tests, this is actually a very, very simple patch. [Build results.](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test_text-5830)

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_12"></a>|@vperic removed owner|
|-|-|

Ok, it just occurred to me that no one actually gave this a "+1" so I'll put it up for review formally.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra set owner to @vperic|
|-|-|

Please merge, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_14"></a>|@vperic set status to `closed`|
|-|-|

(In [35186]) Merge test_text-5830: Make t.test.test_text pass under Python 3

Author: vperic
Reviewers: antoine, thijs
Fixes: [#5830](https://github.com/twisted/twisted/issues/5830)

Don't use the types module in favor of the "type(object) is sometype"
syntagm - types was changed in Python 3 to not include some base types,
like dicts, lists and tuples.
